sentinelhub/download/models.py

```diff
@@ -0,0 +1,257 @@
+
+        request_path = os.path.join(hashed_name, "request.json")
+        response_path = os.path.join(hashed_name, f"response.{self.data_type.extension}")
+
+        return request_path, response_path
+
+    def get_storage_paths(self) -> Tuple[Optional[str], Optional[str]]:
+        """A method that calculates file paths where request payload and response will be saved.
+
+        :return: Returns a pair of file paths, a request payload path and a response path. Each of them can also be
+            `None` if it is not defined.
+        """
+        if self.data_folder is None:
+            return None, None
+
+        request_path, response_path = self.get_relative_paths()
+
+        if request_path is not None:
+            request_path = os.path.join(self.data_folder, request_path)
+        response_path = os.path.join(self.data_folder, response_path)
+
+        self._check_path(response_path)
+        return request_path, response_path
+
+    @staticmethod
+    def _check_path(file_path: str) -> None:
+        """Checks file path and warns about potential problems during saving"""
+        message_problem = None
+        if len(file_path) > 255 and sys_is_windows():
+            message_problem = "File path"
+        elif len(os.path.basename(file_path)) > 255:
+            message_problem = "Filename of"
+
+        if message_problem:
+            message = (
+                f"{message_problem} {file_path} is longer than 255 character which might cause an error while "
+                "saving on disk"
+            )
+            warnings.warn(message, category=SHRuntimeWarning)
+
+
+@dataclass(frozen=True)
+class DownloadResponse:
+    """A class defining a single download response.
+
+    :param request: A download request object for which the response is obtained.
+    :param content: Raw encoded data of the response.
+    :param headers: Headers obtained with the response.
+    :param status_code: Status code of the response.
+    :param elapsed: Number of seconds it took to obtain the response.
+    """
+
+    request: DownloadRequest
+    content: bytes
+    headers: JsonDict = field(default_factory=dict)
+    status_code: Optional[int] = None
+    elapsed: Optional[float] = None
+
+    @classmethod
+    def from_response(cls, response: Response, request: DownloadRequest) -> DownloadResponse:
+        """Creates `DownloadResponse` object from obtained from a service response and request info.
+
+        :param: A service response object.
+        :param: A request for which response was obtained.
+        :return: An instance of a download response object.
+        """
+        return cls(
+            request=request,
+            content=response.content,
+            headers=dict(response.headers),
+            status_code=response.status_code,
+            elapsed=response.elapsed.total_seconds(),
+        )
+
+    @classmethod
+    def from_local(cls, request: DownloadRequest) -> DownloadResponse:
```

Comment:
do we care about the case where the response is saved locally but the `request` object is no longer in python memory? So loading directly from path?

Reply:
The problem is that any instance of `DownloadResponse` object should contain an instance of `DownloadRequest` object as its attribute. This is due to various practical reasons. So loading data directly from a local path would mean we would have to reconstruct the instance of `DownloadRequest` object from the cached content as well. It is possible to implement but atm, I don't see a need to do it.